File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

int solution(vector<int> numbers, int target) {
    int answer = 0;
    queue<pair<int,int>>q;
    q.push({numbers[0],1});
    q.push({numbers[0] * -1,1});
    while(!q.empty()){
        int nx = q.front().first;
        int s = q.front().second;
        q.pop();
        if(s == numbers.size()){
            if(nx == target) answer++;
            continue;
        }
        q.push({nx + numbers[s], s + 1});
        q.push({nx - numbers[s], s + 1});
    }
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public int n, target_copy, answer;
    public int[] numbers_copy;
    
    public void dfs(int idx, int sum){
        if(idx == n){
            if(sum == target_copy) answer++;
            return;
        }
        dfs(idx + 1, sum + numbers_copy[idx]);
        dfs(idx + 1, sum - numbers_copy[idx]);
    }
    
    public int solution(int[] numbers, int target) {
        target_copy = target;
        n = numbers.length;
        numbers_copy = numbers;
        dfs(0,0);
        return answer;
    }
}